torch/jit/__init__.py

```diff
@@ -2034,36 +2035,42 @@ def _check_no_signature(func):
         qual_name = _qualified_name(func)
         raise RuntimeError("Must explicitly add type annotations to overloaded functions: {}".format(qual_name))
 
-def _get_overload_decl_and_defaults(func):
+def _check_no_default_args(func):
+    if get_default_args(func):
+        qual_name = _qualified_name(func)
+        raise RuntimeError("Overloaded default args must be on the implementation function: {}".format(qual_name))
+
+def _get_overload_decl(func):
     _check_no_signature(func)
-    return (torch.jit.get_jit_def(func).decl(), get_default_args(func))
+    _check_no_default_args(func)
+    return torch.jit.get_jit_def(func).decl()
 
 def _get_overloads(obj):
     # check for cached compiled fns
     qual_name = _qualified_name(obj)
     global _compiled_overloaded_fns
     compiled_overloads = _compiled_overloaded_fns.get(qual_name, None)
-    if compiled_overloads is not None:
+    uncompiled_overloads = _jit_internal._get_fn_overloads(qual_name)
+    if compiled_overloads is not None and uncompiled_overloads is None:
```

Comment:
I don't think this is a real use case, but you suggested it as part of the previous PR. https://github.com/pytorch/pytorch/pull/23886#discussion_r312097222

Agreed it would be good to build upon caching logic from other PR.

Reply:
lol. Is this actually allowed to happen (dynamic registration of overloads) or was I being paranoid